cabal-validate/src/OutputUtil.hs

```diff
@@ -0,0 +1,94 @@
+-- | Utilities for printing terminal output.
+module OutputUtil
+  ( printHeader
+  , withTiming
+  ) where
+
+import Control.Exception (catch)
+import System.Console.ANSI
+  ( Color (Cyan, Green, Red)
+  , ColorIntensity (Vivid)
+  , ConsoleIntensity (BoldIntensity)
+  , ConsoleLayer (Foreground)
+  , SGR (Reset, SetColor, SetConsoleIntensity)
+  , setSGRCode
+  )
+import qualified System.Console.Terminal.Size as Terminal
+import System.Process.Typed (ExitCodeException)
+
+import Cli (Opts (..))
+import ClockUtil (diffAbsoluteTime, formatDiffTime, getAbsoluteTime)
+import System.Exit (exitFailure)
+
+-- | Get the width of the current terminal, or 80 if no width can be determined.
+getTerminalWidth :: IO Int
+getTerminalWidth = maybe 80 Terminal.width <$> Terminal.size @Int
+
+-- | Print a header for a given step.
+--
+-- This is colorful and hard to miss in the output.
+printHeader
+  :: String
+  -- ^ Title to print.
+  -> IO ()
+printHeader title = do
+  columns <- getTerminalWidth
+  let left = 3
+      right = columns - length title - left - 2
+      header =
+        setSGRCode [SetConsoleIntensity BoldIntensity, SetColor Foreground Vivid Cyan]
+          <> replicate left '═'
+          <> " "
+          <> title
+          <> " "
+          <> replicate right '═'
+          <> setSGRCode [Reset]
+  putStrLn header
+
+-- | Run an `IO` action and print duration information after it finishes.
+withTiming
+  :: Opts
+  -- ^ @cabal-validate@ options.
+  -> String
+  -- ^ Name for describing the action.
+  --
+  -- Used in a sentence like "@title@ finished after 16.34s".
+  -> IO a
+  -- ^ Action to time.
+  -> IO a
+withTiming opts title action = do
```

Comment:
I was surprised that this guy receives the whole kitchen sink of Opts. If it only uses the initial start time, it may be better to pass exactly that. Otherwise, it reminds me the dreaded DynFlags in GHC and how much pain it took to tear them apart (see the GHC modularity document from a couple of years ago).